NZS3101 design code

Agent6-6-6 opened the pull request:
Hi @robbievanleeuwen 

Still have some minor updates to perform, some more formal verification of results, updating/correcting of docstrings, review of built docs, create some examples, etc..... but the bones are present to perform 3 types of analysis:-

- Normal design analysis: Normal material properties, normal strength reduction factors
- Capacity Protected Element analysis: Normal material properties, strength reduction factor of 1.0
- Overstrength capacity analysis: Likely upper limit material properties (steel strength multiplied by overstrength factor, concrete strength + 15MPa), strength reduction factor of 1.0

- Hope to add a further analysis type in the future to reflect a probable strength analysis (required by guidelines in NZ to assess existing structures)

To achieve the overstrength analysis, the code simply copies the concrete section and updates the material to the modified materials reflecting the likely upper limit material strengths.

There were a few minor changes I made outside of the NZS3101 design class to correct docs, or add other methods required.

Open to any comments you might have on progress!

Agent6-6-6 commented:
Hi @robbievanleeuwen 
I've finished all of the `NZS3101` class code to a working state, implementing all of the features I wanted to add to enable the three types of analysis typically required by NZS3101:2006.

I've also added the ability to assess sections to the NZSEE C5 assessment guidelines (which is basically the bible we need to follow here if assessing existing buildings), there are two additional types of analysis that fit under this umbrella. I wasn't going to do that before the first merge, but isolating at the moment with covid and got nothing better to do!  

- The only feature I'll look to add once it is merged as a separate option would for singly reinforced walls (phi = 0.75 for designing these in plane). 
- Also need to review the default parameters for labels on M/N curve, there is a # TODO somewhere that I need to tidy up.

Still need to do check of built docs and do some basic independent verification of capacities and so forth.

Thanks. 

One thing I've never had anything to do with before is doing the tests, that's the only check failing at the moment. Any advice/help on this would be appreciated?

robbievanleeuwen commented:
Hi @Agent6-6-6, thanks for all your hard work on this! I'll review this weekend :)

Agent6-6-6 commented:
@robbievanleeuwen
Docs should be updated now with c09bdd5efe8cf4e75ea80ebb2a73c62b4137d725, they built fine locally after some tweaking/rearranging. But wasn't sure if there is a way to view the built docs on `*github.io` as part of the pull request tests like you could for `readthedocs` docs to make sure they have all built fine remotely? 

I took the liberty of updating some of the AS3600 docs to be consistent with respect to the material assumptions being in a note.

Essentially the is code ready for review, but do need to do the formal verification of capacities, etc. Hopefully by weekend.

Couple of things to consider:-

- There is quite a long docstring for the `analysis_type` parameter, rather than repeat it 4 times should it just be in the `capacity_reduction_factor` method (since this occurs first) and refer back to that from the other three locations with a link? Will get rid of a lot of repeated code in the 'DesignCode' class?

- In the '__init__' class, the docstring `Inits the NZS3101 class.` is turning up in the docs, note you have the same in the AS3600. Should this just be removed since there are no variables required for the docs? I believe you can filter out certain entries using Sphinx and some custom code in `conf.py` based on a quick web search.

- I note in the AS3600 `moment_interaction_diagram` method that you have no means of feeding a theta to it, you're stuck just doing the analysis at `theta = 0`. Also I wonder if more of the parameters for altering the look of the resultant code design interaction curves should be exposed the user, i.e. missing means of feeding in `theta`, `n_points`, `labels`, etc?

- Should there be a general warning based on the analysis currently being done on constant neutral axis angle vs load angle, and that it may give false impression of demand/capacity when operating close to the curve?

- Need some guidance on tests

- Need to create an example

Any feedback or changes are welcome.

Agent6-6-6 commented:
@robbievanleeuwen having had a little time myself now to look through the examples and how the design code pages are linked together/arranged. I'm just wondering how you want to organise these if more cades are added. In particular these two pages:- 
https://robbievanleeuwen.github.io/concrete-properties/rst/design_codes.html
&
https://robbievanleeuwen.github.io/concrete-properties/notebooks/design_codes.html

I'm wondering if the first link should have the AS3600 & NZS3101 stuff moved on a subpage linked from this page?? Or is the preference just one long list of codes added one after the other?

On the second link, were you thinking something similar but orginaed into subpages with some linked subpages with the actual examples from different codes as they are implemented?? I think that would be the best option to keep things concise?

Anyway f you do get some time this weekend to review, have a think about what your preference might be. If separating them out into separate pages is that something you're able to or prefer to do to prepare for the NZS3101 design code page and example, creating a placeholder page essentially? 

Then I can likely replicate your AS3600 page as a starting point for completing that part of the docs?

robbievanleeuwen commented on concreteproperties/design_codes.py:
```diff
@@ -287,17 +291,23 @@ def create_concrete_material(
-        | - *Flexural tensile strength*: From Cl. 3.1.1.3
+        """Returns a concrete material object to AS 3600:2018.
+
+        .. admonition:: Material assumptions
```
This is nice!

Agent6-6-6 replied:
I thought it stood out more in the docs.

robbievanleeuwen commented on concreteproperties/design_codes.py:
```diff
@@ -727,3 +739,1677 @@ def biaxial_bending_diagram(
+        self, os_design: bool = False, prob_design: bool = False
+    ) -> float:
+        """Function to return the nominal steel reinforcement capacity or the
+            overstrength steel reinforcement capacity of a concrete section.
```
I think there is an unintentional tab here, docs formatting is thrown by this a bit.

Agent6-6-6 replied:
yes missed this, have corrected.

robbievanleeuwen commented:
Hi @Agent6-6-6, great work on this! There is a substantial amount of detail in this and it seems like this will be very useful to NZ engineers! I must admit that a lot of it is over my head, but I think seeing an example (or two) would be helpful in understanding the workings of the NZ code! I'll try to answer all your queries here, let me know if I've missed anything!

1. **Testing**
Unfortunately there is currently no testing coverage for the ``AS3600`` class so nothing to go off. What I intend to do eventually is two fold:
a) Create some unit tests for individual methods, e.g. ``capacity_reduction_factor``, ``get_k_uo`` etc., which have a number of inputs and test for the outputs, covering edge cases. 
b) Find some references with design examples to test a large portion of the module, compare design output from *concreteproperties* with values computed in the reference. Unfortunately since AS3600:2018 is quite fresh I haven't come across any yet, although I haven't searched that hard!

- If you could at least start to implement some tests for some of the individual methods that would be great and of course if you have a test or two for a reference design that would be amazing at verifying that the implementation is working!

2. **Building of Docs**
Unfortunately I don't know of a way of previewing the docs on github pages for pull requests. However if you have all the dependencies installed locally, your locally built docs should be identical to what is generated on github pages so building the docs locally is the way to go. Readthedocs is a bit more of a black box and this is one of the reasons (amongst others) that I've moved away from them, but they do have some nice features which I'll miss!

3. **Long docstring for the ``analysis_type`` parameter**
I always think it's a good idea not to repeat yourself with code, so yes this could be either appear once in another method. Or (perhaps preferably) this could be moved to the NZS design code docs page and the doc string could just say refer to docs, with a link to a sub-heading in the NZS docs page.

4. **init docstring**
Yes I have these enabled by default as most classes have parameters here. I'm not too fussed with including this in the API docs, however this can be easily configured to be hidden when you write the NZS docs page.

5. **AS3600 Moment Interaction Method**
Yes this is an oversight by me, thanks! I'll get this cleaned up.

6. **Load axis angle warning**
Yes good idea, I will put this on the Design Codes 'landing page' in the docs.

7. **Docs Structure**
I think the structure should be as follows:

- Design Codes Page (i.e. https://robbievanleeuwen.github.io/concrete-properties/rst/design_codes.html): general overview of design codes in *concreteproperties*, with load axis angle warning and links to the following:
  - AS3600 page
  - NZS3101 page
  - Other design code pages
- Examples page (i.e. https://robbievanleeuwen.github.io/concrete-properties/rst/examples.html)
  - Examples 1 to 8 to remain as is
  - Rename example 9 to AS3600 Design Code
  - Add example 10, NZS3101 Design Code

I will open issues above where relevant and modify the docs structure to suit item 7.

I think it would also be a good idea to change the location of the design codes files. I propose the following:

- concreteproperties
  - design_codes
    - design_code.py (Abstract design code)
    - as3600.py (AS3600 design code)
    - nzs3101.py (NZS3101 design code)

I will make the above changes, I think if everything sits in one file it will very quickly get out of hand!

Agent6-6-6 commented on concreteproperties/design_codes.py:
```diff
@@ -727,3 +739,1677 @@ def biaxial_bending_diagram(
+
+    def create_steel_material(
+        self,
+        steel_grade: str = None,
```
Happy to do this but I'm not following the logic for this change I think, as what I had makes it optional in the code and docs, and sets the desired default??

What is the additional benefit of this approach, or what is it doing relative to what is currently noted (noting you do the exact same thing in the `AS3600` class when assigning optional defaults?)
![image](https://user-images.githubusercontent.com/36808164/191070665-5aef1114-024a-472c-b9e0-fa43a8bcf715.png)

Agent6-6-6 replied:
Actually, this is a rare occasion where it is picking it up, but generally it is not picking up the optional parameters as optional in the docs, assume that's why this is required throughout?

robbievanleeuwen replied:
Have a look at the python docs for type hints [here](https://docs.python.org/3/library/typing.html#typing.Optional). There is a difference between providing a default value for the argument and having an argument that is optional, i.e. ``= None``.

Agent6-6-6 replied:
Yeah updated this to correctly use the `Optional` type

Agent6-6-6 commented on concreteproperties/design_codes.py:
```diff
@@ -727,3 +739,1677 @@ def biaxial_bending_diagram(
+        """
+
+        name: str
+        steel_grade: str
```
noted this is an oversight, have added and rather than repeat the longer docstring associated with this I've directed to the `create_steel_material()` method.

Agent6-6-6 commented on concreteproperties/material.py:
```diff
@@ -39,6 +39,8 @@ class Concrete(Material):
     :param flexural_tensile_strength: Absolute value of the concrete flexural
         tensile strength
     :param colour: Colour of the material for rendering
+    :param meshed: If set to True, the entire material region is meshed; if set to
```
Noted, will remove (and other instances). I thought it had been inadvertently missed!

Agent6-6-6 commented on concreteproperties/material.py:
```diff
@@ -74,6 +76,8 @@ class Steel(Material):
     :param density: Steel density (mass per unit volume)
     :param stress_strain_profile: Steel stress-strain profile
     :param colour: Colour of the material for rendering
+    :param meshed: If set to True, the entire material region is meshed; if set to
```
as above response, will remove

Agent6-6-6 commented on concreteproperties/material.py:
```diff
@@ -92,6 +96,8 @@ class SteelBar(Steel):
     :param density: Steel bar density (mass per unit volume)
     :param stress_strain_profile: Steel bar stress-strain profile
     :param colour: Colour of the material for rendering
+    :param meshed: If set to True, the entire material region is meshed; if set to
```
as above response, will remove

Agent6-6-6 commented on concreteproperties/design_codes.py:
```diff
@@ -727,3 +739,1677 @@ def biaxial_bending_diagram(
+            assessment guidelines C5.4.3
+        :param stress_strain_profile: Steel bar stress-strain profile
+        :param colour: Colour of the material for rendering
+        :param meshed: If set to True, the entire material region is meshed; if set to
```
Noted, assuming you mean always equal to `false` for the `SteelBar` material.
Have removed.

Agent6-6-6 commented:
Hi @robbievanleeuwen I've chucked the beginnings of a design example within the notebooks folder in the docs, still needs quite a bit of finessing but hopefully it demonstrates how the different types of analysis work. Refer 2e2e8cb452517bd1eaad91b1d5c791959bba870e

1. **Testing** - my total experience with tests and creating tests is currently non-existent, so I'm probably going to need some guidance or steering in the right direction. 

- Is there currently some similar tests that might be helpful to try work through converting to cover the NZS3101 code that are part of `concrete-properties` or even `section-properties` already, the sort of thing that gives me the general idea so I can take a crack at it and go from there?

2. Noted.

3. Yeah I'll look at moving it to the docs page once I develop this. For now I'll just remove it from all but one method and we can see how it sits best there or on docs page once that is created.

4. Noted, I still have to write this page, once you've separated out the AS3600 page as noted I'll develop that design code page off the basis of what you've currently done for AS3600.

5. Noted. I'll just update NZS3101 similar method to match however you implement this if any different to maintain consistency.

6. Noted. I think it's worth pointing out how it works currently, and that it is intended in the future at some point to move to a load angle based approach.

7. Noted, that docs structure all makes sense, and is in line with what I was anticipating. Makes it easier to expand to other codes in the future which is nice. 

Agree with separating it out into separate `NZS3101.py` files under `design_codes` subfolder. I'll make the change at my end this week sometime.

I'm also looking at implementing the Mander unconfined/confined concrete stress-strain relationship, as it is utilised for moment curvature analyses in the NZSEE C5 assessment guidelines.

Addressed majority of the review comments in b17982d916eb864715e70cbc7f5e9074ad5ea630. Just have to review what I was doing with the `get_fracture_strain` method!

robbievanleeuwen commented:
Hi @Agent6-6-6,

**Testing** - there are example tests in the ``concreteproperties/tests`` folder. ``test_reinforced_concrete_basics.py`` covers some design examples, ``test_stress_strain_profile.py`` covers some basic unit tests and ``test_rotation.py`` uses ``parametrize`` to rotate a section through a number of angles to check that the results are consistent with different neutral axis angles. *sectionproperties* also has a number of tests in a similar directory, so this could be a starting point.

robbievanleeuwen commented:
Hi @Agent6-6-6, I modified the file structure of the package and docs as agreed with 12de46d.

Agent6-6-6 commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -688,6 +688,16 @@ def get_yield_strength(
 
         return self.yield_strength
 
+    def get_fracture_strain(
```
Yes I think this works, provided `min(self.strains)` is always equal to the fracture strain. 

There is currently no built in steel stress/strain profiles where this is not the case. If it turns out to be the case in the future for any reason then can re-add `get_fracture_strain()` method to base class `StressStrainProfile` returning `self.fracture_strain`, and for concrete a similar one that returns `self.ultimate_strain`. 

That way you're returning the exact values specified in generating the stress/strain relationship which is what I was aiming for as I'm using those values to generate new modified properties suitable for different analysis types vs from the generated stress/strain relationship. There is a subtle difference potentially depending on future stress/strain profiles and whether or not `max(self.strains) == self.ultimate_strain` & `min(self.strains) == self.fracture_strain`.

Unless you think they should just be added now?

Agent6-6-6 replied:
Actually to avoid any future issues, I think I'll update the code to simply use `__getattribute__()` to grab the `stress_strain_profile` or `ultimate_stress_strain_profile` stored variables for `fracture_strain`, `ultimate_strain`, `yield_strength`, `compressive_strength`, etc... to get the exact values used to generate the materials and stress/strian profiles. 

I think this is a better approach as not all of these variables have provided functions to retrieve directly, often only via the generated stresses and strains which may not be the same as the value specified to actually generate that profile (i.e. in the case of confined concrete models as noted in #47 i.e. `max(self.strains) != self.ultimate_strain`. 

Will resolve on this basis @robbievanleeuwen.

Agent6-6-6 replied:
Addressed with dbe6e350751d300996170ca0703ec34de71617e3

Agent6-6-6 commented:
@robbievanleeuwen, wondering if you are able to take a look at the NZS3101 notebook example, I am getting an error in the moment-curvature analysis part of the example when finding the newly added failure curvature feature from #61.
`ValueError: f(a) and f(b) must have different signs`

robbievanleeuwen commented:
@Agent6-6-6 I got it to work by putting a small horizontal section on the end of the concrete stress-strain profile:

```python
concrete_25_prob.stress_strain_profile.strains = np.append(concrete_25_prob.stress_strain_profile.strains, 4.01e-3)
concrete_25_prob.stress_strain_profile.stresses = np.append(concrete_25_prob.stress_strain_profile.stresses,23.45255773)
```

In the algorithm there was probabaly a neutral axis depth that resulted in a large compressive strain, causing the stress-strain profile extrapolation to give a tensile stress - this probably caused the error.

Agent6-6-6 commented:
Sweet, I'll update the ModifiedMander class to include this for interpolation purposes.

Agent6-6-6 commented:
Hi @robbievanleeuwen 

I think I'm almost at completion here..... a couple of things to do related to some of the current pull requests that have not been merged yet:-

- [ ] Need to make a few changes to `moment_interaction_diagram()` when #66 is merged to add the analysis limits and so forth and additional control points.
- [ ] Need to review once #62 is merged, but it should all be working fine, current tests are all based on your pull request so pass locally but fail on commit until this is merged.
- [ ] As noted in #62 in my 1st comment, the way the compression block accounts for the bar holes is I believe throwing out the neutral axis depth slightly compared to checking by hand methods, I've commented out the offending test for now which is done at an axial load equal to zero. I do get the same moment capacity though with the change to the geometric centroid. At higher axial loads I get the same answer for neutral axis depth which is a bit weird and needs a bit further investigation.
- [ ] Need to write a few more tests to cover some of my more recent additions and any code currently not touched by the tests. Code coverage check won't complete as tests fail as per above, so cannot fully complete this until the above two pull requests are merged and I get back into a position where the tests and code coverage is working 👍.
- [ ] Need to add the `modifiedmander` stress-strain relationship to `materials.rst` in docs. Note I had set up the docstring with some generated matplotlib images to explain the various parameters for this profile (generated via the `mander_images.py` file). I was wondering if the same should be done for the other stress-strain profiles to explain the relevant parameters and give a generalised pretty picture of the stress/strain relationship. These images come through on the `materials.rst` file as well as part of the docstring before providing the small code examples. Anyway have a review and see what you think, but I think it makes it really clear what the stress and strain parameters are for example which may not be fully apparent by just reviewing the code-generated plots from the short examples. 

Thoughts on adding similar for each stress-strain profile?
i.e.

![mander_unconfined](https://user-images.githubusercontent.com/36808164/201446286-1ddf2184-acb3-46f3-a427-0e46ba5763e3.png)
![mander_confined](https://user-images.githubusercontent.com/36808164/201446288-a81fba87-ab2f-40c1-9cf9-2b80ac614f1c.png)

- [ ] Final review of docs/example for any errors/omissions


Once the above is done I'll mark it for a final review! But if you have any comments in the interim feel free to add some review comments!

Thanks

robbievanleeuwen commented:
Hi @Agent6-6-6,

Thanks for your work on this, it's looking really excellent!

- re: #66, I'll work to get that merged ASAP. If not today, it will probably be Thursday or Friday
- re: #62, are you happy for this to be merged as is?
- re: pretty stress-strain plots - I think these look really great and think they would be a great addition! Perhaps we could even change the default ``StressStrainProfile.plot_stress_strain()`` so that they all have this style? When it comes to including these in the docs (e.g. in ``materials.rst``), I would much prefer if the images were auto-generated rather than importing image files. Have a look at the way I do it at the moment using the ``.. plot`` directive. There are a number of benefits to this e.g. code changes are automatically captured, there is no required post-processing, smaller/cleaner repo and there is consistency between all images, especially with multiple contributors who might be on different systems/running different versions of ``matplotlib``. I believe if we rework ``StressStrainProfile.plot_stress_strain()`` to include the style changes and maybe add another method for each stress-strain profile that generates the above plot (e.g. ``StressStrainProfile.plot_definition()``) which we then put in the plot directive, this could work really nicely!

Agent6-6-6 commented:
Re #62, OK by me to merge, I think the differences in results neutral axis depth are just in the way you've worked the bars into the concrete as holes. But it does raise the question on what values to put on the tests and how to independantly check the concrete-properties neutral axis depth I guess if accepting that for the test. Can investigate further after merge to determine exactly why there are the differences so a note can go into the docs or something to explain. 

Regarding images - Yeah I just wasn't sure how you did the plot on building docs, and didn't necessarily want the images changing each time and requiring a tweak of the formatting or positioning of labels, etc. As the intent was they were fairly generic just to get across the various parameters in a more obvious manner than just describing in text. There are quite a few modifications to get the images looking correctly proportioned in the docs, plot size, text scale, manually plotting axes, custom tick marks, manual axes 'arrows', etc. But I'll have a look and see how you did the auto-generated plot images and see if it works, as was imagining you'd have all the code to generate them in another file due to the customisations. 

I agree adding that style or something similar as another option would be good for reports, etc. Although if printing multiple stress-strain plots on one figure, not sure how good it will turn out with the fill.

Agent6-6-6 commented:
@robbievanleeuwen just wanted to confirm the following and your thoughts based on your prior post regarding pretty stress-strain plots and the way you envisaged these being organised for the docs if generating the plots on build:-

If we add in `stress_strain_profile.py` another class (something like `StressStrainPlots` with individual functions to generate each custom plot for the docs, is that basically what you mean? Then within each stress-strain profile class docstring call like 

```
.. plot::
      :include-source: False

      from concreteproperties.stress_strain_profiles import StressStrainPlots as ssplt

      ssplt.mander_unconfined_plot()
```


Or do we include the pretty plot for the docs as a sub-function within each individual stress-strain profile class? and call like:-
```
.. plot::
      :include-source: False

      self.mander_unconfined_plot()
```
I feel like both these ways gets pretty messy, and fills the package folder with a lot of untestable code?

So, I wonder if a separate class located in another folder/file like `stress_strain_profile_plots.py` in the docs folder is a better approach to keep it all in one common location for ease of maintenance and to keep it away from the main concreteproperties folder? And have methods in this class for each plot image you want to modify and tweak for the pretty image displayed in the docs.
```
from docs.source.doc_plots.stress_strain_profile_plots import StressStrainPlots as ssplt

      ssplt.mander_unconfined_plot()
```
Thoughts?  

One problem I've come up against that I have not been able to solve yet is the `render=False` in a method to generate these pretty plots from the docs does not seem to be respected when it is called from the docs. So I cannot modify the plot before using plt.show(). If you have a look at what I was doing in mander_images.py to see what I was doing previously to generate the *.png images, if you have any idea on this as well?

Agent6-6-6 commented:
@robbievanleeuwen, ignore the above, I've gotten it all working now using the below method locating individual files within a new docs `_static/doc_plots` folder to build each plot .... it was an exercise in frustration I can tell you..... !  (see c29df771bcd29c9edb60a22d718ef036f7db1a38) 

```
    .. plot:: ./_static/doc_plots/mander_unconfined_plot.py mander_unconfined_plot
      :include-source: False
      :caption: ModifiedMander Parameters for Unconfined Concrete

    .. plot:: ./_static/doc_plots/mander_confined_plot.py mander_confined_plot
      :include-source: False
      :caption: ModifiedMander Parameters for Confined Concrete
```

Suggest once the class is merged if you have another preferred location for these files in the package structure then they can always be moved. But to get it to work you can only have a single plot per file. Otherwise, it just repeats the same plot multiple times if you have multiple different methods within one file (see above point on frustration!). 

I'll look to generate others for all of the stress-strain profiles images along with showing the parameters as a separate pull request at some point once #40 is merged.

Just a few tests to go on the ModifiedMander class and then I think I'm actually done and it should be ready for review (finally)!

Agent6-6-6 commented:
@robbievanleeuwen, marked ready for review. Let me know if you have any comments. Otherwise, merge if happy!

robbievanleeuwen commented:
@Agent6-6-6 thanks this looks great! My kid's got chickenpox at the moment, but hopefully I'll have a chance to have a look at this at the end of this week/early next week! Love your work!

robbievanleeuwen commented:
Hi @Agent6-6-6, finally got the time to go through the entire PR. I started writing a review then realised it would be easiest if I commit some minor changes as a lot of it was styling, linting things etc. We can revert anything you don't agree with, but here's the gist:

- I use the VS code import organiser, which just keeps things neat and consistent
- Docstrings with backslashes can cause issues if they aren't raw strings, i.e. add ``r`` before the ``"""``
- Ensuring all methods have a return type
- Ensuring all variables are bound before returning (these probably don't need a ``raises ValueError`` in the docstring as they are generally private methods)
- Type checking fixes

Other than the above everything looks really fantastic! Thank you so much for taking the time to make the code readable and understandable, and the documentation is amazing!

I have a few comments regarding the ``ModifiedMander`` which I'll address with some code comments. Other than that, I think we're good to go!

Agent6-6-6 commented:
@robbievanleeuwen, yeah go for it.

Agree with items 1 to 3, thought I had import organiser installed but apparently not. 

Item 4 no idea what you're really talking about 😊 similar for 5. So feel free to amend as required and I'll have a look see if it triggers any other minor amendments or tidying up.

robbievanleeuwen commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+        E_sec = f_cc / eps_cc
+
+        if self.conc_confined:
+            self.strains = np.linspace(0, eps_c_max, self.n_points)
```
Should be of type ``List[float]`` not a numpy array, suggest adding ``.tolist()`` to end of statement.

I.e. ``self.strains = np.linspace(0, eps_c_max, self.n_points).tolist()``

Agent6-6-6 replied:
need to convert to list later on as `x = self.strains / eps_cc` throws error otherwise. Think my thinking was easier to operate on np array than add loop to calculate x.

Will update to convert to list after this calculation and amend code below this to use `.insert` & `.append` on the lists as intent was a list is returned for the stresses and strains as per other methods.

Agent6-6-6 replied:
See commit bc11ce88828df3cfcc572478fe8aa87d542ededd

robbievanleeuwen commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+        if self.conc_confined:
+            self.strains = np.linspace(0, eps_c_max, self.n_points)
+            # add eps_cc point corresponding to max stress point at end
+            self.strains = np.append(self.strains, eps_cc)
```
As above

Agent6-6-6 replied:
as per above

robbievanleeuwen commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+            # add eps_cc point corresponding to max stress point at end
+            self.strains = np.append(self.strains, eps_cc)
+        else:
+            self.strains = np.linspace(0, min(2 * eps_cc, eps_c_max), self.n_points)
```
As above

Agent6-6-6 replied:
as per above

robbievanleeuwen commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+        else:
+            self.strains = np.linspace(0, min(2 * eps_cc, eps_c_max), self.n_points)
+            # add eps_cc point corresponding to max stress point at end
+            self.strains = np.append(self.strains, eps_cc)
```
As above

Agent6-6-6 replied:
as per above

robbievanleeuwen commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+
+        # add spalling branch if specified for unconfined curve
+        if not self.conc_confined and self.conc_spalling:
+            self.strains = np.append(self.strains, self.eps_sp)
```
As above, ``self.strains.append(self.eps_sp)``

Agent6-6-6 replied:
as per above

robbievanleeuwen commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+        # add spalling branch if specified for unconfined curve
+        if not self.conc_confined and self.conc_spalling:
+            self.strains = np.append(self.strains, self.eps_sp)
+            self.stresses = np.append(self.stresses, 0)
```
As above

Agent6-6-6 replied:
as per above

robbievanleeuwen commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+
+        if self.conc_tension:
+            # add tension stress/strain limit
+            self.strains = np.insert(self.strains, 0, -eps_t)
```
As above, ``self.strains.insert(0, -eps_t)``.

Applies to all below.

Agent6-6-6 replied:
as per above

robbievanleeuwen commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+        self.ultimate_strain = max(self.strains)
+
+        # add small horizontal compressive strain to improve interpolation
+        self.strains = np.append(self.strains, self.strains[-1] + 1e-12)
```
As above.

Agent6-6-6 replied:
as per above

See commit bc11ce88828df3cfcc572478fe8aa87d542ededd for revised code to address `ModifiedMander` review coments, converting numpy array to list as soon as practical so output type is List as intended.

Agent6-6-6 commented:
For item 4, I think (at least some of) the input was checked/validated earlier in the code so by the time you get to that location there is no way you could have a wrong value. So returning a value error seemed redundant. But appreciate if you called that method by itself with a wrong input it would fail. But I really need to refresh my mind what I did there as I could be wrong in the way I addressed this. But I guess I'll just need to raise new tests against these new lines to get back to full coverage to close it out if it makes code more complete with no loopholes. 

Will review the comments in ModifiedMander class. Cannot recall why I just kept it as numpy array throughout, will review if it works as a list.

Agent6-6-6 commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+
+        if self.conc_confined and self.sect_type in ["circ_hoop", "circ_spiral"]:
+            self.b = 0
+            self.w_dash = 0
```
Yes, strictly speaking I guess it should. Setting the variable to zero is simply removing this variable from the next check on validating inputs by setting it to something that is not `None`. Will change.

Agent6-6-6 commented on concreteproperties/stress_strain_profile.py:
```diff
@@ -457,6 +457,367 @@ def __post_init__(
+            i
+            for i in self.__dataclass_fields__.keys()
+            if self.__getattribute__(i) is None
+        ]
```
Yeah basically, if any of the required variables are not provided by the user for the 'sect_type' specified when defining a confined concrete stress-strain profile, then the analysis just defaults to an unconfined concrete model and gives a warning and provides a list of the variables that were not defined if the user wishes to resolve.

Agent6-6-6 commented:
Hi @robbievanleeuwen 

For `assign_analysis_section()` added a new test for the additional `else` statement you added to check for an incorrect `analysis_type`. Note though this incorrect `analysis_type` is checked earlier in the code when it is first encountered within the `capacity_reduction_factor()` method so it is effectively redundant in the `assign_analysis_section()` method if you're using the NZS3101 class as intended. But added for completeness. See commit d3c450421e3994a12024cc33cecb18e680947cdf.

Started working on the others but I got to thinking that the same redundant check looks like it applies to all the other `else` statements added. At the point where the methods execute there is no way that the wrong value could be assigned as it has been checked earlier in the code execution at the first instance it was required in other methods and was not being checked at each and every subsequent instance it was used within subsequent methods. 

I can add tests to cover the additional `else` statements and amend some of the value error messages to be the same as the first location where it is checked to be consistent throughout if it is really required, but it would only ever be executed in the tests so it seems like adding code for no real purpose.

You may have missed one `else` as well in `concrete_capacity() method` for after `elif self.section_type.lower() in ["wall", "wall_sr_s", "wall_sr_m"]:`? But this still is a redundant check that would never be executed except for within a test as it is checked earlier in the `assign_concrete_section()` method the first time it is encountered for example. 

Please review given the redundant nature of these if you still actually want them included (I mean no harm in having them, but they are not required if you're using the `NZS3101` class as intended by design. But I can see if someone is accessing a method outside of the Class you'd want to catch the failures, but it was never written with this use case in mind as everything you need is intended to be in the `NZS3101` class, from creating section to analysing it. But I guess until people start using it hard to say how people might use it and if this would therefore throw an error). 

Note there were tests previously to capture the first instance of a wrong input for `analysis_type` & `section_type` for example, showing that these subsequent methods would never be executed with the wrong inputs. There was no way you could create a section consistent with the NZS3101 class without utilising the `SteelbarNZ` class for the reinforcement. 

Some tests to hit these redundant `else` statements are a bit of a pain as multiple methods strung together would require the same `else` statement (`max_comp_strength()` calls  `concrete_capacity()` for example both would require testing of `section_type`, both of which would require an `else` statement to capture the redundant incorrect input for completeness).

Anyway just wanted to double-check if this is really required prior to going to the trouble of adding additional `else` statements which are never hit by the actual code execution and then writing specific tests to hit these lines of code for the purposes of only the code coverage percentage? As that seems a bit redundant to me writing code/tests for something that is never able to be executed in a normal use case due to the preceding code by design checking the same thing.

Let me know your thoughts.

robbievanleeuwen commented:
Hi @Agent6-6-6, I am happy with the code as is, thanks for these changes. We definitely don't need to aim for 100% test coverage, I don't believe in writing tests after the fact just to get full coverage, although I am definitely guilty of not writing enough tests myself! Having said that, your ~97% coverage is commendable!

I have built the docs locally and everything looks fantastic. I just have to do a few fixes with the new shapely release in sectionproperties, hence the <2 requirement for now. If you are happy with the code in it's current state I will merge and create a new release!

Agent6-6-6 commented:
There are a couple of the statements you added that I'll just update to be similar valueerror message to other locations to be consistent. But apart from that I didn't have anything else. I'll do this later today when I'm free to spend some time on it.

Agent6-6-6 commented:
Hi @robbievanleeuwen,

Nearly complete now with the following changes being committed in cbef839f5ef856f0db6f7f33351bd3eff564d842 to tidy up the loose ends:-

- Changed one of the additional `ValueError` checks from `max_comp_strength()` to `concrete_capacity()` as the `concrete_capacity()` is called from `max_comp_strength()`, so this is the initial location that should throw the `ValueError` if either method is called independently. Added a test for this.
- Added tests for the non `SteelBarNZ` material `ValueError`, as realised it is quite possible that you could assemble a section and assign it to the `NZS3101` class as `concrete_section` by bypassing the `create_concrete_section()` method and end up with another steel material that is not compatible with the `NZS3101` class requirements. However, I'm not hitting these lines of code with the test which I don't yet understand so need to check further....
- Updated docstrings for completeness to add the `:raises: ValueError` explanations where the new checks were located.

Just need to look into the second point further then it should be ready to merge.

Agent6-6-6 commented:
@robbievanleeuwen, all good to merge now with a1656b32354229a6ec5d19ba94a9edaf5b86fdbe addressing point above.
